README.md
This is a content security policy violation collector written in Golang.

It has been designed to listen on port 8080 and accept POST payloads
containing the violation report. It captures the report and will write
it to STDOUT via Go's logger.

A neat little feature of this tool is that it automatically ignores
unactionable reports. Check out the [default list][1] if you're interested.

### Installation

```sh
$ go get github.com/jacobbednarz/go-csp-collector
```

Alternatively, you can download the binaries from the [release page][2].

### Running

```sh
$ go build main.go -p csp-collector
$ ./csp-collector
```

### Endpoints

- `POST /`: accepts a CSP violation report (recommended to use `/csp` for future proofing though).
- `POST /csp`: accepts a CSP violation report.
- `POST /csp/report-only`: same as `/csp` but appends a `report-only` attribute to the log line. Helpful if you have enforced and report only violations and wish to separate them.

#### Building for Docker

You will either need to build within a docker container for the purpose, or use `CGO_ENABLED=0` flag
to make the build compatible with alpine linux in a docker container.

```sh
$ CGO_ENABLED=0 go build main.go -p csp-collector
```

### Command Line Options

| Flag                    | Description                                                                                                                                                                                       |
| ----------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| version                 | Shows the version string before exiting                                                                                                                                                           |
| debug                   | Runs in debug mode producing more verbose output                                                                                                                                                  |
| port                    | Port to run on, default 8080                                                                                                                                                                      |
| filter-file             | Reads the blocked URI filter list from the specified file. Note one filter per line                                                                                                               |
| health-check-path       | Sets path for health checkers to use, default \/\_healthcheck                                                                                                                                     |
| log-client-ip           | Include a field in the log with the IP delivering the report, or the value of the `X-Forwarded-For` header, if present.                                                                           |
| log-truncated-client-ip | Include a field in the log with the truncated IP (to /24 for IPv4, /64 for IPv6) delivering the report, or the value of the `X-Forwarded-For` header, if present. Conflicts with `log-client-ip`. |
| truncate-query-fragment | Remove all query strings and fragments (if set) from all URLs transmitted by the client                                                                                                           |
| query-params-metadata   | Log all query parameters of the report URL as a map in the `metadata` field                                                                                                                       |

See the `sample.filterlist.txt` file as an example of the filter list in a file

### Request metadata

Additional information can be attached to each report by adding a `metadata`
url parameter to each report. That value will be copied verbatim into the
logged report.

For example a report sent to `https://collector.example.com/?metadata=foobar`
will include field `metadata` with value `foobar`.

If `query-params-metadata` is set, instead all query parameters are logged as a
map, e.g. `https://collector.example.com/?env=production&mode=enforce` will
result in `"metadata": {"env": "production", "mode": "enforce"}` in JSON
format, and `metadata="map[env:production mode:enforce]"` in default format.

### `report-only` mode

If you'd like to recieve report only violations on a different URL

### Output formats

The output format can be controlled by passing `--output-format <type>`
to the executable. Available formats are:

- **Text**: A key/value output that quotes all the values. Example:
  `blocked_uri="about:blank" ...`
- **JSON**: Single line, compressed JSON object. Example:
  `{"blocked_uri":"about:blank"}`

The default formatter is text.

### Writing to a file instead of just STDOUT

If you'd rather have these violations end up in a file, I suggest just
redirecting the output into a file like so:

```sh
$ ./csp-collector 2>> /path/to/violations.log
```

### Visualisation

This project purposely doesn't try to solve the problem of visualing the
violation data because there are already a bunch of great solutions out
there. Once you have your violations being collected, be sure to slurp
them into your favourite log aggregation tool.

### Deployments

Currently supported deployment mechanisms:

- [kubernetes/helm][3]
- [systemd][4]

[1]: https://github.com/jacobbednarz/go-csp-collector/blob/master/sample.filterlist.txt
[2]: https://github.com/jacobbednarz/go-csp-collector/releases
[3]: https://github.com/jacobbednarz/go-csp-collector/tree/master/deployments/kubernetes-helm/README.md
[4]: https://github.com/jacobbednarz/go-csp-collector/tree/master/init


CHANGELOG.md
## v0.0.14 (Unreleased)

- Bump to Go 1.21

## v0.0.13 (Unreleased)

## v0.0.12 

- Shuffle internals around to add dedicated CSP endpoints and make way for NEL and reporting API.

## v0.0.11

**Improvements**

- Add `log-client-ip` flag for logging client IP with requests
- Add `log-truncated-client-ip` flag to use in conjunction with `log-client-ip` for logging only the IP range
- Add `truncated-query-fragment` flag to remove all query strings and fragments (if set) from all URLs transmitted by the client
- Add `query-params-metadata` flaf to log all query parameters of the report URL as a map in the `metadata` field
- Bump to Go 1.19

## v0.0.10

**Improvements**

- Fix README links for kubernetes-helm documentation

## v0.0.9 - Definitely Maybe

**Features**

- Add support for Kubernetes Helm deployments ([#17](https://github.com/jacobbednarz/go-csp-collector/issues/17))

**Improvements**

- Filter MS application schemes ([#43](https://github.com/jacobbednarz/go-csp-collector/issues/43))

## v0.0.8

**Improvements**

- Updated format of `-version` output to be `MAJOR.MINOR.PATCH+GIT_SHA` ([#42](https://github.com/jacobbednarz/go-csp-collector/issues/42))

## v0.0.7

**Improvements**

- Allow metadata smuggling by appending `metadata` query parameter to reports ([#40](https://github.com/jacobbednarz/go-csp-collector/issues/40))

## v0.0.6

**Improvements**

- Ignore document-uri values that don't start with `http` ([#39](https://github.com/jacobbednarz/go-csp-collector/issues/39))

## v0.0.5

**Improvements**

- Adds ability to have comments in filter list file ([#37](https://github.com/jacobbednarz/go-csp-collector/issues/37))

## v0.0.4

**Improvements**

- Dependency upgrades
- Update Go version and use modules

**Features**

- Add support for JSON output
- Docker support
- More invalid URLs

## v0.0.3

**Improvements**

- Update README links and releases mention
- Adds coverage for ensuring all required keys are outputted
- Add coverage for ValidateViolation function
- Use lowercase in error messages because the cool kids do
- Don't track the dist directory
- 👋 1.6 and hello 1.8
- Add coverage for healthcheck endpoint
- Don't include hostname in urlStr param
- Add test coverage for disallowed methods
- Add test coverage for disallowed methods on root

**Fixes**

- Don't panic for invalid payloads
- Ensure we always have a timestamp

## v0.0.2

**Features**

- Add `-debug` flag
- add darwin builds

**Improvements**

- add goreleaser for packaging


csp_collector.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"net/http"
	"net/netip"
	"os"
	"strconv"
	"strings"

	log "github.com/sirupsen/logrus"
)

// CSPReport is the structure of the HTTP payload the system receives.
type CSPReport struct {
	Body CSPReportBody `json:"csp-report"`
}

// CSPReportBody contains the fields that are nested within the
// violation report.
type CSPReportBody struct {
	DocumentURI        string      `json:"document-uri"`
	Referrer           string      `json:"referrer"`
	BlockedURI         string      `json:"blocked-uri"`
	ViolatedDirective  string      `json:"violated-directive"`
	EffectiveDirective string      `json:"effective-directive"`
	OriginalPolicy     string      `json:"original-policy"`
	Disposition        string      `json:"disposition"`
	ScriptSample       string      `json:"script-sample"`
	StatusCode         interface{} `json:"status-code"`
}

const (
	// Default health check url.
	defaultHealthCheckPath = "/_healthcheck"
)

var (
	// Rev is set at build time and holds the revision that the package
	// was created at.
	Rev = "dev"

	// Shared defaults for the logger output. This ensures that we are
	// using the same keys for the `FieldKey` values across both formatters.
	logFieldMapDefaults = log.FieldMap{
		log.FieldKeyTime:  "timestamp",
		log.FieldKeyLevel: "level",
		log.FieldKeyMsg:   "message",
	}

	// Default URI Filter list.
	defaultIgnoredBlockedURIs = []string{
		"resource://",
		"chromenull://",
		"chrome-extension://",
		"safari-extension://",
		"mxjscall://",
		"webviewprogressproxy://",
		"res://",
		"mx://",
		"safari-resource://",
		"chromeinvoke://",
		"chromeinvokeimmediate://",
		"mbinit://",
		"opera://",
		"ms-appx://",
		"ms-appx-web://",
		"localhost",
		"127.0.0.1",
		"none://",
		"about:blank",
		"android-webview",
		"ms-browser-extension",
		"wvjbscheme://__wvjb_queue_message__",
		"nativebaiduhd://adblock",
		"bdvideo://error",
	}
)

func init() {
	log.SetOutput(os.Stdout)
	log.SetLevel(log.InfoLevel)
}

func trimEmptyAndComments(s []string) []string {
	var r []string
	for _, str := range s {
		if str == "" {
			continue
		}

		// ignore comments
		if strings.HasPrefix(str, "#") {
			continue
		}

		r = append(r, str)
	}
	return r
}

func main() {
	version := flag.Bool("version", false, "Display the version")
	debugFlag := flag.Bool("debug", false, "Output additional logging for debugging")
	outputFormat := flag.String("output-format", "text", "Define how the violation reports are formatted for output.\nDefaults to 'text'. Valid options are 'text' or 'json'")
	blockedURIFile := flag.String("filter-file", "", "Blocked URI Filter file")
	listenPort := flag.Int("port", 8080, "Port to listen on")
	healthCheckPath := flag.String("health-check-path", defaultHealthCheckPath, "Health checker path")
	truncateQueryStringFragment := flag.Bool("truncate-query-fragment", false, "Truncate query string and fragment from document-uri, referrer and blocked-uri before logging (to reduce chances of accidentally logging sensitive data)")

	logClientIP := flag.Bool("log-client-ip", false, "Log the reporting client IP address")
	logTruncatedClientIP := flag.Bool("log-truncated-client-ip", false, "Log the truncated client IP address (IPv4: /24, IPv6: /64")

	metadataObject := flag.Bool("query-params-metadata", false, "Write query parameters of the report URI as JSON object under metadata instead of the single metadata string")

	flag.Parse()

	if *version {
		fmt.Printf("csp-collector (%s)\n", Rev)
		os.Exit(0)
	}

	if *debugFlag {
		log.SetLevel(log.DebugLevel)
	}

	if *outputFormat == "json" {
		log.SetFormatter(&log.JSONFormatter{
			FieldMap: logFieldMapDefaults,
		})
	} else {
		log.SetFormatter(&log.TextFormatter{
			FullTimestamp:          true,
			DisableLevelTruncation: true,
			QuoteEmptyFields:       true,
			DisableColors:          true,
			FieldMap:               logFieldMapDefaults,
		})
	}

	log.Debug("Starting up...")
	ignoredBlockedURIs := defaultIgnoredBlockedURIs
	if *blockedURIFile != "" {
		log.Debugf("Using Filter list from file at: %s\n", *blockedURIFile)

		content, err := os.ReadFile(*blockedURIFile)
		if err != nil {
			log.Fatalf("Error reading Blocked File list: %s", *blockedURIFile)
		}
		ignoredBlockedURIs = trimEmptyAndComments(strings.Split(string(content), "\n"))
	} else {
		log.Debug("Using Filter list from internal list")
	}

	log.Debugf("Blocked URI List: %s", ignoredBlockedURIs)
	log.Debugf("Listening on TCP Port: %s", strconv.Itoa(*listenPort))

	http.HandleFunc(*healthCheckPath, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}

		w.WriteHeader(http.StatusOK)
	})

	http.Handle("/", &violationReportHandler{
		blockedURIs:                 ignoredBlockedURIs,
		truncateQueryStringFragment: *truncateQueryStringFragment,

		logClientIP:          *logClientIP,
		logTruncatedClientIP: *logTruncatedClientIP,
		metadataObject:       *metadataObject,
	})
	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", strconv.Itoa(*listenPort)), nil))
}

type violationReportHandler struct {
	truncateQueryStringFragment bool
	blockedURIs                 []string

	logClientIP          bool
	logTruncatedClientIP bool
	metadataObject       bool
}

func (vrh *violationReportHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		log.WithFields(log.Fields{
			"http_method": r.Method,
		}).Debug("Received invalid HTTP method")

		return
	}

	decoder := json.NewDecoder(r.Body)
	var report CSPReport

	err := decoder.Decode(&report)
	if err != nil {
		w.WriteHeader(http.StatusUnprocessableEntity)
		log.Debugf("Unable to decode invalid JSON payload: %s", err)
		return
	}
	defer r.Body.Close()

	reportValidation := vrh.validateViolation(report)
	if reportValidation != nil {
		http.Error(w, reportValidation.Error(), http.StatusBadRequest)
		log.Debugf("Received invalid payload: %s", reportValidation.Error())
		return
	}

	var metadata interface{}
	if vrh.metadataObject {
		metadataMap := make(map[string]string)
		query := r.URL.Query()

		for k, v := range query {
			metadataMap[k] = v[0]
		}

		metadata = metadataMap
	} else {
		metadatas, gotMetadata := r.URL.Query()["metadata"]
		if gotMetadata {
			metadata = metadatas[0]
		}
	}

	lf := log.Fields{
		"document_uri":        report.Body.DocumentURI,
		"referrer":            report.Body.Referrer,
		"blocked_uri":         report.Body.BlockedURI,
		"violated_directive":  report.Body.ViolatedDirective,
		"effective_directive": report.Body.EffectiveDirective,
		"original_policy":     report.Body.OriginalPolicy,
		"disposition":         report.Body.Disposition,
		"script_sample":       report.Body.ScriptSample,
		"status_code":         report.Body.StatusCode,
		"metadata":            metadata,
		"path":                r.URL.Path,
	}

	if vrh.truncateQueryStringFragment {
		lf["document_uri"] = truncateQueryStringFragment(report.Body.DocumentURI)
		lf["referrer"] = truncateQueryStringFragment(report.Body.Referrer)
		lf["blocked_uri"] = truncateQueryStringFragment(report.Body.BlockedURI)
	}

	if vrh.logClientIP {
		ip, err := getClientIP(r)
		if err != nil {
			log.Warnf("unable to parse client ip: %s", err)
		}
		lf["client_ip"] = ip.String()
	}

	if vrh.logTruncatedClientIP {
		ip, err := getClientIP(r)
		if err != nil {
			log.Warnf("unable to parse client ip: %s", err)
		}
		lf["client_ip"] = truncateClientIP(ip)
	}

	log.WithFields(lf).Info()
}

func (vrh *violationReportHandler) validateViolation(r CSPReport) error {
	for _, value := range vrh.blockedURIs {
		if strings.HasPrefix(r.Body.BlockedURI, value) {
			err := fmt.Errorf("blocked URI ('%s') is an invalid resource", value)
			return err
		}
	}

	if !strings.HasPrefix(r.Body.DocumentURI, "http") {
		return fmt.Errorf("document URI ('%s') is invalid", r.Body.DocumentURI)
	}

	return nil
}

func truncateQueryStringFragment(uri string) string {
	idx := strings.IndexAny(uri, "#?")
	if idx != -1 {
		return uri[:idx]
	}

	return uri
}

func truncateClientIP(addr netip.Addr) string {
	// Ignoring the error is statically safe, as there are always enough bits.
	if addr.Is4() {
		p, _ := addr.Prefix(24)
		return p.String()
	}

	if addr.Is6() {
		p, _ := addr.Prefix(64)
		return p.String()
	}

	return "unknown-address"
}

func getClientIP(r *http.Request) (netip.Addr, error) {
	if s := r.Header.Get("X-Forwarded-For"); s != "" {
		addr, err := netip.ParseAddr(s)
		if err != nil {
			return netip.Addr{}, fmt.Errorf("unable to parse address from X-Forwarded-For=%s: %w", s, err)
		}

		return addr, nil
	}

	addrp, err := netip.ParseAddrPort(r.RemoteAddr)
	if err != nil {
		return netip.Addr{}, fmt.Errorf("unable to parse remote address %s: %w", r.RemoteAddr, err)
	}

	return addrp.Addr(), nil
}
